zcash_client_backend/src/data_api/chain.rs

```diff
@@ -199,7 +201,9 @@ pub trait BlockSource {
     /// `from_height`, applying the provided callback to each block. If `from_height`
     /// is `None` then scanning will begin at the first available block.
     ///
-    /// Returns an error if `from_height` is set to a block that does not exist in `self`.
+    /// Returns the range of block heights for blocks passed to `with_block` if any blocks were
+    /// available, `Ok(None)` if no blocks were provided, or an error if `from_height` is set to a
+    /// block that does not exist in `self`.
```

Comment:
This reads weirdly to me, but I can't think of a nicer way to state it. "No blocks provided" and "blocks passed" are the phrases that are tripping me up. I think what it's trying to convey is that the range contains all blocks that `with_block` was called on, and `Ok(None)` means that `with_block` was never called?

Separately, what should this return if `from_height` is `Some(h)` and `limit` is `Some(0)`? I think that semantically, this should return `Ok(Some(Range(h, h)))` (i.e. the empty range starting at `h`), even though `with_block` was never called, but the answer to this somewhat depends on how the return value is documented (per above). The empty range is not fully duplicative of `Ok(None)`, due to `from_height` being optional (an empty range still has a defined start height, and `from_height` being `None` is documented as "use an undefined start height" - although we could improve that definition by returning the empty range starting at Sapling activation, for instance).

Reply:
I struggled here with whether I should return `Option<ScanSummary>` or just `ScanSummary`, but my conclusion was that I couldn't actually produce a useful `ScanSummary` if no height information was provided and the `BlockSource` didn't pick up anything.

Reply:
I realized that it actually _is_ possible to provide a non-Optional `ScanSummary`, because `from_height` is now non-optional.